core/debug/src/debug_box.c

```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
 
 uint32_t g_debug_interrupt_sp[UVISOR_MAX_BOXES];
 
+void debug_print(const uint8_t * message_buffer, uint32_t size)
+{
+	/* Abort printing if debug box wasn't initialized yet. */
+	if (!g_debug_box.initialized) {
+		return;
+	}
+
+	/* Debug print relies on "deprivilege and return" flow that will currently
+	 * only work if:
+	 * - executed with exceptions enabled
+     * - executed from within the exception context, i.e. when IPSR isn't 0
+     * - the current exception priority is lower than of Hard Fault
+	 * These limitation are dictated by the fact that SVC is used for
+	 * deprivileging and in order for it to work SVC must be executed with
+	 * exceptions enabled and its priority must be higher than of the current exception.
+	 * SVC's priority is set to be higher than of any other exception with
+	 * programmable priority but this still can't beat NMI and Hard Fault.
+	 */
+
+	/* Make sure we're within an exception and exception are enabled. */
+	int exception = (int)(__get_IPSR() & 0x1FF);
+	if (!exception || __get_PRIMASK() || __get_FAULTMASK())	{
+		return;
+	}
+
+	/* Bring exception number to CMSIS format and check exception's priority. */
+	exception -= NVIC_OFFSET;
+	if (exception == NonMaskableInt_IRQn || exception == HardFault_IRQn) {
+		return;
+	}
+
+	/* Place the message on the interrupts stack allocating
+	 * the size rounded up to the next value divisible by 4. */
```

Comment:
Changed the comment to:

/* Place the message on the interrupt stack allocating the size rounded up
 * to the next value divisible by 4. This way the stack stays aligned on
 * word boundary. */

Reply:
Why does the stack need to be aligned on 4-byte boundary? Sometimes it has to be aligned to 8-byte boundary, right? Is 4-byte alignment always enough here?

Reply:
As far as I know 4-byte alignment is an architectural requirement - you can't even set the two lowest bits of SP to be non-zero. 8-byte alignment is enforced by the architecture by automatically inserting padding when creating and exception stack frame.

In our case 4-byte alignment is sufficient, context_forge_exc_sf() will add padding if necessary.

Reply:
AAPCS compliance requires double-word stack alignment.
Since we are calling functions implemented in C  - we better to be complaint with calling convention.
@Patater - nice catch!

Reply:
I've changed the alignment to be 8 bytes.